src/Components/Components/src/Rendering/ComponentState.cs

```diff
@@ -222,5 +227,31 @@ private void DisposeBuffers()
             ((IDisposable)CurrentRenderTree).Dispose();
             _latestDirectParametersSnapshot?.Dispose();
         }
+
+        internal Task DisposeInBatchAsync(RenderBatchBuilder batchBuilder)
+        {
+            _componentWasDisposed = true;
+
+            CleanupComponentStateResources(batchBuilder);
+
+            try
+            {
+                var result = ((IAsyncDisposable)Component).DisposeAsync();
+                if (result.IsCompletedSuccessfully)
+                {
+                    return Task.CompletedTask;
+                }
+                else
+                {
+                    return Awaited(result);
```

Comment:
Is this any different from `result.AsTask()`? There's probably some subtlety I'm missing. If so, would you consider adding a comment to clarify what we're doing the await for?

Reply:
Can the entire try/catch be replaced with `return ((IAsyncDisposable)Component).DisposeAsync().AsTask();`? I don't know what it is adding.

Reply:
I suspect one of the intentions here is to ensure it never throws synchronously (to simplify things for the caller), which may explain some of it, but @javiercn can better clarify the rest!

Reply:
AsTask probably works, I'm just too used to task and I didn't think of just calling AsTask, I would check though to ensure it does the right thing.

As for the try..catch, we want to avoid the caller having to deal with capturing exceptions, this is a result of us not using async to avoid the state machine. We follow this pattern in several places.

Reply:
From what I can see, https://github.com/dotnet/runtime/blob/bf2e135c12cbd34aeba2fa4a31d0e84184041a17/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs#L191-L228 AsTask will do just this, so I think I'll replace the entire thing with that.

@JamesNK thanks for the suggestion!

Reply:
Nevermind, I still need the try..catch for when the thing throws synchronously

Reply:
Actually, AsTask doesn't work here because when the exception is thrown synchronously the result is never set, which leads to an incorrect result.

Reply:
It can be used in the async code path, so I've removed the Awaited bit

Reply:
LGTM!